tests/importer/importer.cpp

```diff
@@ -0,0 +1,761 @@
+
+    auto importer = libcellml::Importer::create();
+    auto parser = libcellml::Parser::create();
+    auto printer = libcellml::Printer::create();
+    auto model = parser->parseModel(fileContents("importer/trunk.cellml"));
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_FALSE(model->hasUnresolvedImports());
+    EXPECT_EQ(size_t(2), importer->libraryCount());
+    auto flat = importer->flattenModel(model);
+    EXPECT_EQ(flatModelString, printer->printModel(flat));
+}
+
+TEST(Importer, importFilesWithSameName)
+{
+    // This test is to see whether the importer correctly resolves imported sources, where
+    // the imported files all have the same file name.
+
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("importer/model.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto importer = libcellml::Importer::create();
+
+    // Manually add a model to the library with the "model.cellml" key.
+    auto anotherModel = libcellml::Model::create("anotherModelWithKey_model.cellml");
+    anotherModel->addComponent(libcellml::Component::create("b"));
+    anotherModel->addUnits(libcellml::Units::create("a"));
+    importer->addModel(anotherModel, "model.cellml");
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(0), importer->issueCount());
+
+    // 4 items in the library.
+    EXPECT_EQ(size_t(4), importer->libraryCount());
+    // 3 of them are resolved from external locations -> have absolute URLs as keys.
+    EXPECT_EQ(size_t(3), importer->externalDependencyCount());
+
+    EXPECT_FALSE(model->hasUnresolvedImports());
+
+    // Compare the library items against their model name to be sure they're different.
+    std::vector<std::string> modelNames = {"layer3_model", "layer2_model", "layer1_model", "anotherModelWithKey_model.cellml"};
+    for (size_t i = 0; i < importer->libraryCount(); ++i) {
+        EXPECT_EQ(modelNames[i], importer->library(i)->name());
+    }
+}
+
+TEST(Importer, coverageNestedImportsClearingModels)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("importer/nested_components.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    auto importer = libcellml::Importer::create();
+    importer->resolveImports(model, resourcePath("importer/"));
+
+    EXPECT_EQ(size_t(0), importer->issueCount());
+
+    EXPECT_FALSE(model->hasUnresolvedImports());
+    importer->clearImports(model);
+    EXPECT_TRUE(model->hasUnresolvedImports());
+}
+
+TEST(Importer, importSourceGetSetModel)
+{
+    auto model = libcellml::Model::create("m");
+    auto component = libcellml::Component::create("c");
+    auto importsource = libcellml::ImportSource::create();
+
+    component->setImportSource(importsource);
+    EXPECT_EQ(nullptr, importsource->model());
+
+    importsource->setModel(nullptr);
+    EXPECT_EQ(nullptr, importsource->model());
```

Comment:
Why are those two lines needed? The `importSource`'s model is already null, so...? I would remove those two lines.

Reply:
It's to check that the "clearing" of the model via the set(null) kind of thing works as expected.  The weak vs strong pointer thing made it less than straightforward.

Reply:
I appreciate and that's why I am happy with lines 766 and 767. I just don't understand why it's also done at lines 753 and 754.